File: app/src/main/java/com/checkin/app/checkin/Shop/ShopRepository.java
package com.checkin.app.checkin.Shop;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.BaseRepository;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.checkin.app.checkin.Misc.GenericDetailModel;
import com.checkin.app.checkin.Shop.Private.MemberModel;
import com.checkin.app.checkin.Shop.ShopJoin.ShopJoinModel;
import com.checkin.app.checkin.Utility.ProgressRequestBody;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;

/**
 * Created by Bhavik Patel on 24/08/2018.
 */

public class ShopRepository extends BaseRepository {
    private static ShopRepository INSTANCE;
    private final WebApiService mWebService;

    private ShopRepository(Context context) {
        mWebService = ApiClient.Companion.getApiService(context);
    }

    public static ShopRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (ShopRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ShopRepository(application.getApplicationContext());
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<GenericDetailModel>> registerShop(ShopJoinModel model) {
        return new NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<GenericDetailModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postRegisterShop(model));
            }

            @Override
            protected void saveCallResult(GenericDetailModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> updateShopDetails(final RestaurantModel restaurantModel) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.putRestaurantManageDetails(restaurantModel.getPk(), restaurantModel));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> updateShopContact(final long shopId, final ObjectNode data) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.putRestaurantContactVerify(shopId, data));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<RestaurantModel>> getShopModel(long shopId) {
        return new NetworkBoundResource<RestaurantModel, RestaurantModel>() {

            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<RestaurantModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantDetails(shopId));
            }

            @Override
            protected void saveCallResult(RestaurantModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<RestaurantModel>> getShopManageModel(long shopId) {
        return new NetworkBoundResource<RestaurantModel, RestaurantModel>() {

            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<RestaurantModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantManageDetails(shopId));
            }

            @Override
            protected void saveCallResult(RestaurantModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<RestaurantModel>>> getShops() {
        return new NetworkBoundResource<List<RestaurantModel>, List<RestaurantModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<RestaurantModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurants());
            }

            @Override
            protected void saveCallResult(List<RestaurantModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<MemberModel>>> getRestaurantMembers(long shopId) {
        return new NetworkBoundResource<List<MemberModel>, List<MemberModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<MemberModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantMembers(shopId));
            }

            @Override
            protected void saveCallResult(List<MemberModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> addRestaurantMember(long shopId, MemberModel data) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.postRestaurantMember(shopId, data));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> updateRestaurantMember(long shopId, MemberModel shopMember) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.putRestaurantMember(shopId, shopMember.getUserId(), shopMember));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> removeRestaurantMember(long shopId, long userId) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.deleteRestaurantMember(shopId, userId));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> deleteRestaurantCover(long shopId, int index) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.deleteRestaurantCover(shopId, index));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {

            }
        }.getAsLiveData();
    }

    public Call<GenericDetailModel> postRestaurantLogo(long mShopPk, File pic, ProgressRequestBody.UploadCallbacks listener) {
        RequestBody requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic);
        ProgressRequestBody requestBody = new ProgressRequestBody(requestFile, listener);
        final MultipartBody.Part body = MultipartBody.Part.createFormData("logo", "cover.jpg", requestBody);
        return mWebService.postRestaurantLogo(mShopPk, body);
    }

    public Call<GenericDetailModel> postRestaurantCover(long mShopPk, int index, File pic, ProgressRequestBody.UploadCallbacks listener) {
        RequestBody requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic);
        ProgressRequestBody requestBody = new ProgressRequestBody(requestFile, listener);
        final MultipartBody.Part body = MultipartBody.Part.createFormData("image", "cover.jpg", requestBody);
        return mWebService.postRestaurantCover(mShopPk, index, body);
    }
}

File: app/src/main/java/com/checkin/app/checkin/Shop/ShopRepository.kt
package com.checkin.app.checkin.Shop

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.Shop.Private.MemberModel
import com.checkin.app.checkin.Shop.ShopJoin.ShopJoinModel
import com.checkin.app.checkin.data.BaseRepository
import com.checkin.app.checkin.data.network.ApiClient
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.home.model.NearbyRestaurantModel
import com.checkin.app.checkin.misc.models.GenericDetailModel
import com.checkin.app.checkin.utility.ProgressRequestBody
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import java.io.File


class ShopRepository private constructor(context: Context) : BaseRepository() {
    private val mWebService: WebApiService = ApiClient.getApiService(context)

    fun nearbyRestaurants(id: Int): LiveData<Resource<List<NearbyRestaurantModel>>> =
            object : NetworkBoundResource<List<NearbyRestaurantModel>, List<NearbyRestaurantModel>>() {
                override fun shouldUseLocalDb(): Boolean {
                    return false
                }

                override fun createCall(): LiveData<ApiResponse<List<NearbyRestaurantModel>>> {
                    return RetrofitLiveData(mWebService.nearbyRestaurants(if (id == 0) null else id))
                }
            }.asLiveData

    fun registerShop(model: ShopJoinModel): LiveData<Resource<GenericDetailModel>> {
        return object : NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<GenericDetailModel>> {
                return RetrofitLiveData(mWebService.postRegisterShop(model))
            }
        }.asLiveData
    }

    fun updateShopDetails(restaurantModel: RestaurantModel): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.putRestaurantManageDetails(restaurantModel.getPk(), restaurantModel))
            }
        }.asLiveData
    }

    fun updateShopContact(shopId: Long, data: ObjectNode): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.putRestaurantContactVerify(shopId, data))
            }
        }.asLiveData
    }

    fun getShopModel(shopId: Long): LiveData<Resource<RestaurantModel>> {
        return object : NetworkBoundResource<RestaurantModel, RestaurantModel>() {

            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<RestaurantModel>> {
                return RetrofitLiveData(mWebService.getRestaurantDetails(shopId))
            }
        }.asLiveData
    }

    fun getShopManageModel(shopId: Long): LiveData<Resource<RestaurantModel>> {
        return object : NetworkBoundResource<RestaurantModel, RestaurantModel>() {

            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<RestaurantModel>> {
                return RetrofitLiveData(mWebService.getRestaurantManageDetails(shopId))
            }
        }.asLiveData
    }

    fun getRestaurantMembers(shopId: Long): LiveData<Resource<List<MemberModel>>> {
        return object : NetworkBoundResource<List<MemberModel>, List<MemberModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<MemberModel>>> {
                return RetrofitLiveData(mWebService.getRestaurantMembers(shopId))
            }
        }.asLiveData
    }

    fun addRestaurantMember(shopId: Long, data: MemberModel): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.postRestaurantMember(shopId, data))
            }
        }.asLiveData
    }

    fun updateRestaurantMember(shopId: Long, shopMember: MemberModel): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.putRestaurantMember(shopId, shopMember.userId, shopMember))
            }
        }.asLiveData
    }

    fun removeRestaurantMember(shopId: Long, userId: Long): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.deleteRestaurantMember(shopId, userId))
            }
        }.asLiveData
    }

    fun deleteRestaurantCover(shopId: Long, index: Int): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.deleteRestaurantCover(shopId, index))
            }
        }.asLiveData
    }

    fun postRestaurantLogo(mShopPk: Long, pic: File, listener: ProgressRequestBody.UploadCallbacks<GenericDetailModel>): Call<GenericDetailModel> {
        val requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic)
        val requestBody = ProgressRequestBody(requestFile, listener)
        val body = MultipartBody.Part.createFormData("logo", "cover.jpg", requestBody)
        return mWebService.postRestaurantLogo(mShopPk, body)
    }

    fun postRestaurantCover(mShopPk: Long, index: Int, pic: File, listener: ProgressRequestBody.UploadCallbacks<GenericDetailModel>): Call<GenericDetailModel> {
        val requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic)
        val requestBody = ProgressRequestBody(requestFile, listener)
        val body = MultipartBody.Part.createFormData("image", "cover.jpg", requestBody)
        return mWebService.postRestaurantCover(mShopPk, index, body)
    }

    companion object : SingletonHolder<ShopRepository, Application>({ ShopRepository(it.applicationContext) })
}
